### PR TITLE
[ENG-2278] Add file/folder delete to <Files::widget>

### PR DESCRIPTION
**Pull request by fabmiz**
- Ticket: [ENG-2278]
- Feature flag: n/a

## Purpose
Add file/folder delete to the files widget

- [x] Add file/folder delete to the files widget
- [x] Language warning users about the 5GB limit
- [x] Update language `You may attach up to 5 file(s) to this question. You may attach files... to account for no-project workflows`
- [x] Language for file/folder delete confirmation modal.

~Waiting on product for language updates.~
Screen Shot 2021-02-02 at 5 17 29 PM

## Summary of Changes
- Update `<DeleteButton />` to not propagate click by default.
- Add file/folder delete button to `<Files::Item />`
- Add a file/folder delete task to `<Files::Manager />`
- Update `<Files::Manager >` to use `onDeleteFile` and allow consumers of the file widget to dictate what happens
upon successful delete in this case `<Registries/SchemaBlockRenderer::Editable />`
- Add `onDeleteFile` action to `<Registries/SchemaBlockRenderer::Editable::Files />` 
- Add acceptance tests

## QA Notes
Regression test for files widget on draft registrations.

[ENG-2278]: https://openscience.atlassian.net/browse/ENG-2278

### CODE REVIEW

File: lib/osf-components/addon/components/files/manager/component.ts [hunk 1]
```diff
@@ -198,6 +200,28 @@ export default class FilesManagerComponent extends Component {
...
+    deleteFileTask = task(function *(this: FilesManagerComponent, file: File) {
+        try {
+            yield file.delete();
+
+            if (this.onDeleteFile) {
+                this.onDeleteFile(
+                    file,
+                    { callback: file.isFolder ? this.goToParentFolder.bind(this, file) : null },
+                );
+            }
```
Comment (brianjgeiger) [inline lib/osf-components/addon/components/files/manager/component.ts:213]: Out of curiosity, why is this necessary? Rather than, say, just calling `goToParentFolder()` if this is a folder.

Comment (brianjgeiger) [reply lib/osf-components/addon/components/files/manager/component.ts:213]: Can you delete a folder that you're not currently in? What happens then? It seems like this will always bump you up one level regardless of whether that's appropriate or not.

File: lib/osf-components/addon/components/registries/schema-block-renderer/editable/files/component.ts [hunk 2]
```diff
@@ -99,4 +99,17 @@ export default class Files extends Component {
     onAddFile(addedFile: File) {
         this.select(addedFile);
     }
+
+    @action
+    onDeleteFile(deletedFile: File, options: { callback?: () => void }) {
+        if (deletedFile.isFolder) {
```
Comment (brianjgeiger) [inline lib/osf-components/addon/components/registries/schema-block-renderer/editable/files/component.ts:105]: We're already checking if the item is a folder from when we add the callback. Are you sure that we'd never want a callback if the deletedFile is a file? I don't know any cases off the top of my head, but the double-check seems a bit limiting.

**Review by brianjgeiger — COMMENTED** [review]
I have a couple of questions around the delete folder callback, because I don't fully understand the complexity. Overall it looks great.

File: lib/osf-components/addon/components/registries/schema-block-renderer/editable/files/component.ts (hunk 2 above)
Comment (fabmiz) [reply lib/osf-components/addon/components/registries/schema-block-renderer/editable/files/component.ts:105]: Good catch. It makes more sense to check against callback.

File: lib/osf-components/addon/components/files/manager/component.ts (hunk 1 above)
Comment (fabmiz) [reply lib/osf-components/addon/components/files/manager/component.ts:213]: Currently, we only support deleting a folder you're currently in. The idea is that it's safer for the user to know the contents of the folder before deleting it. Regarding `goToParentFolder()`, the idea is to keep the files widget generic so it's easier to extend in the future. and because `goToParentFolder` is the most logical thing to do after folder deletion it's passed in as a callback to the consumer of the `<FilesWidget />`.

File: translations/en-us.yml [hunk 3]
```diff
@@ -1567,6 +1567,7 @@ osf-components:
             editButtonAriaLabel: 'Edit {questionLabel}'
         schema-block-renderer/editable/files:
             instructions: 'You may attach up to 5 file(s) to this question. You may attach files that you already have in OSF Storage in this <a href={nodeUrl}>{projectOrComponent}</a> or upload (drag and drop) a new file from your computer. Uploaded files will automatically be added to this <a href={nodeUrl}>{projectOrComponent}</a> so that they can be registered. To attach files from other components or an add-on, first add them to this <a href={nodeUrl}>{projectOrComponent}</a>.'
+            instructionsNoProject: 'You may attach up to 5 file(s) to this question. Files cannot total over 5GB in size.<br><br>Uploaded files will automatically be archived in this registration. They will also be added to a related project that will be created for this registration.'
```
Comment (fabmiz) [inline translations/en-us.yml:1570]: Will be used once we have `DraftNode`

File: lib/osf-components/addon/components/registries/schema-block-renderer/editable/files/component.ts [hunk 4]
```diff
@@ -99,4 +99,24 @@ export default class Files extends Component {
     onAddFile(addedFile: File) {
         this.select(addedFile);
     }
+
+    @action
+    onDeleteFile(deleteFileOrFolder: File, options: { callback?: () => void }) {
+        let filesToUnselect: File[] = [];
+        if (deleteFileOrFolder.isFolder) {
+            filesToUnselect = this.selectedFiles
+                .filter(({ materializedPath }) => materializedPath.includes(deleteFileOrFolder.materializedPath));
```
Comment (fabmiz) [inline lib/osf-components/addon/components/registries/schema-block-renderer/editable/files/component.ts:108]: added to unselect folder descendant files upon deletion

**Review by brianjgeiger — APPROVED** [review]

File: lib/osf-components/addon/components/files/item/template.hbs [hunk 5]
```diff
@@ -11,6 +11,20 @@
...
+        <span local-class='delete-button'>
+            <DeleteButton
+                data-test-delete-current-folder='{{this.item.id}}'
+                @buttonLabel={{t 'osf-components.files-widget.delete'}}
+                @delete={{fn @filesManager.deleteFileTask this.item}}
+                @small={{true}}
+                @noBackground={{true}}
+                @modalTitle={{t 'osf-components.files-widget.confirm_delete.title_folder' filename=this.item.itemName}}
+                @confirmButtonText={{t 'osf-components.files-widget.delete'}}
+                @shouldStopPropagation={{true}}
```
Comment (adlius) [inline lib/osf-components/addon/components/files/item/template.hbs:23]: What would happen if we don't stop the propagation of the event here?

Comment (fabmiz) [reply lib/osf-components/addon/components/files/item/template.hbs:23]: There is a [click listener](https://github.com/CenterForOpenScience/ember-osf-web/pull/1136/files#diff-c8d8238e29580b496a64ab96c2ed2687e221725552bf33eae293792bee506835L6) on the parent div, which would also handle the `delete-button` click resulting, in this case, in the file being `selected` or `unselected`

**Review by adlius — APPROVED** [review]
